Add Local Custodian Codes

baisa opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Added Rake task responsible for populating local_custodian_code column with data from the CSV file. `import_from_csv` method is implemented in `local_custodian_codes_importer` helper.

[Trello card](https://trello.com/c/OQKm9z4V/2758-add-local-custodian-codes-to-local-links-manager-5)

baisa committed:
Add local_custodian_code column.

This column is added to the Local Authorities table.

ChrisBAshton commented on spec/lib/local-links-manager/import/local_custodian_codes_importer_spec.rb:
```diff
@@ -0,0 +1,42 @@
+      let!(:local_authority_3) { create(:local_authority, slug: "stockton-on-tees", local_custodian_code: nil) }
+
+      it "populates code with value in different formats" do
+        LocalLinksManager::Import::LocalCustodianCodesImporter.new.import_from_csv("spec/lib/local-links-manager/import/fixtures/addressbase-local-custodian-codes-test.csv")
```
This can be refactored slightly:

```suggestion
        described_class.new.import_from_csv("spec/lib/local-links-manager/import/fixtures/addressbase-local-custodian-codes-test.csv")
```

ChrisBAshton commented on spec/lib/local-links-manager/import/local_custodian_codes_importer_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    context "local custodian codes for Local Authorities in Northern Ireland" do
+      let!(:local_authority_1) { create(:local_authority, slug: "belfast", local_custodian_code: nil) }
+
+      it "populates code with value" do
```
As a general rule, we want to avoid using `context` when it only wraps around a single test. It's a sign that it's been slightly over-engineered. These tests would read more clearly as something like:

```ruby
it "extracts local custodian codes from CSVs containing Northern Ireland Local Authorities" do
  local_authority = create(:local_authority, slug: "belfast", local_custodian_code: nil)
  described_class.new.import_from_csv("spec/lib/local-links-manager/import/fixtures/addressbase-local-custodian-code-Northern-Ireland-test.csv")

  expect(local_authority_1.reload.local_custodian_code).to eq "8132"
end
```

ChrisBAshton commented on spec/lib/tasks/local_custodian_codes_from_csv_spec.rb:
```diff
@@ -0,0 +1,10 @@
+      args = Rake::TaskArguments.new(%i[filename], ["spec/lib/local-links-manager/import/fixtures/addressbase-local-custodian-codes-test.csv"])
+      local_authority = create(:local_authority, slug: "leicestershire", local_custodian_code: nil)
+      expect { Rake::Task["import:local_custodian_codes"].execute(args) }
+        .to change { local_authority.reload.local_custodian_code }.from(nil).to("2460")
```
This is kind of a repeat of the unit tests in local_custodian_codes_importer_spec.rb. We've already tested the functionality extensively in there, so this test can simply check that a new instance of `LocalLinksManager::Import::LocalCustodianCodesImporter` was created.

It might be tricky to do that without using `allow_any_instance_of` - [some ideas](https://github.com/alphagov/locations-api/pull/14#discussion_r732869828) here.

baisa replied:
Thank you for looking at it @ChrisBAshton.  In this case, I think it's better to check the actual change than mocks. The test looks the same, but it has a different role. This is more a feature/acceptance spec, as it checks one of the application entrypoints, while test for LocalCustodianCodeImporter is a simple unit spec. I would avoid to use mocks when they are not necessary, leaving them mostly for unit specs, where they help test classes in isolation, like in the PR you linked.

ChrisBAshton replied:
Fair enough - thanks for looking 👍

ChrisBAshton commented on lib/tasks/import/local_custodian_codes_from_csv.rake:
```diff
@@ -0,0 +1,6 @@
+namespace :import do
```
I think the rake task and associated tests can be added in a commit after the `LocalCustodianCodesImporter` has been created. Nice self-contained commits that layer functionality on - makes it easier to review and track changes.

baisa replied:
Thank you for noticing this one, I will keep it in mind for the future, but perhaps will leave it in one commit this time as our docs advise to avoid rewriting commits unless really necessary?

ChrisBAshton replied:
Are you referring to [these docs](https://gds-way.cloudapps.digital/standards/source-code.html#branching-merging-conventions)? I can see where it says that, but it's referring to commits that have already been pushed to `main`. Commits on unmerged branches, like this one, should "make liberal use of Git’s history rewriting features" 👍 

I agree that doc is a bit ambiguous. In this case though, we know that we "are very sure that no-one else will be affected by you rewriting the branch history", as there are no branches dependent on this one. I might raise a follow-up PR to the GDS Way to try and clarify this a bit (or you're welcome to!)

baisa replied:
I have split the commit @ChrisBAshton . Happy to raise follow-up PR 👍

ChrisBAshton commented on addressbase-local-custodian-codes-street.csv:
```diff
@@ -0,0 +1,393 @@
+﻿Local Custodian Code,Authority,Region,Country
```
Similarly, I think the CSVs can be added in their own self-contained commit. The commit would ideally describe how you created these CSVs too, or where you downloaded them from.

I also think they should be grouped somewhere, e.g. in a `data/` folder instead of at the root of the repository.

baisa replied:
Good point @ChrisBAshton, I have moved them to data folder in the separate commit.

ChrisBAshton replied:
Thanks @baisa - much easier to review now!

It's worth adding a [link to the Trello card](https://trello.com/c/OQKm9z4V/2758-add-local-custodian-codes-to-local-links-manager-5) in the commit message itself, rather than just refer to "the card", as it saves a developer having to find the PR associated with the commit, to find the Trello card.

baisa replied:
@ChrisBAshton makes sense! I have added the link.

ChrisBAshton reviewed (commented):
Nice one, @baisa - this is looking great! Just a few comments below.

baisa committed:
Add CSV files.

This files for Great Britain, Northern Ireland and Islands were provided with the card(https://trello.com/c/OQKm9z4V/2758-add-local-custodian-codes-to-local-links-manager-5) and can be found in data folder

ChrisBAshton reviewed (approved):
Thanks @baisa , the PR is looking great! I've left a couple of comments on the conversations above, which you may want to look at before merging, but they're non-blocking comments, so you can merge if you want.

One last (incredibly nitpicky!) comment: I've noticed the commit message summaries all end in a full stop `.`.  We don't usually do this. Not a big deal at all, but probably worth bearing in mind for future work, to keep the style of our commit messages consistent as an organisation.

baisa committed:
Add LocalCustodianCodesImporter helper

Helper provides a method that will update local_custodian_code value. Added tests

baisa committed:
Add Rake task.

This task is responsible for calling LocalCustodianCodesImporter. Added test

ChrisBAshton reviewed (approved):
Perfect - thanks!
